packages/devtools_app/lib/src/memory/memory_heap_tree_view.dart

```diff
@@ -643,62 +736,37 @@ class HeapTreeViewState extends State<HeapTree> with AutoDisposeMixin {
     );
   }
 
-  void updateListOfSnapshotsUnderAnalysisNode() {
-    final AnalysesReference analysesNode = findAnalysesNode(controller);
-    assert(analysesNode != null);
-
-    for (final snapshot in controller.snapshots) {
-      final currentSnapDT = snapshot.collectedTimestamp;
-
-      if (analysesNode.children.length == 1) {
-        final node = analysesNode.children.first;
-        if (node is AnalysisReference && node.name.isEmpty) {
-          setState(() {
-            analysesNode.collapse();
-            analysesNode.children.clear();
-          });
-        }
-      }
-
-      AnalysisSnapshotReference analyzeSnapshot;
-      final foundAnalysis = controller.completedAnalyses
-          .where((analysis) => analysis.dateTime == currentSnapDT);
-      if (foundAnalysis.isNotEmpty) {
-        // If there's an analysis of this snapshot then show it.
-        analyzeSnapshot = foundAnalysis.single;
-      } else {
-        analyzeSnapshot = AnalysisSnapshotReference(currentSnapDT);
-      }
-
-      analysesNode.addChild(analyzeSnapshot);
-    }
-  }
-
   void _analyze() {
-    final AnalysesReference analysesNode = findAnalysesNode(controller);
+    final AnalysesReference analysesNode = controller.findAnalysesNode();
     assert(analysesNode != null);
 
-    final currentSnapDT = controller.snapshots.last.collectedTimestamp;
-
-    // If an analysis of the current snapshot exist then do nothing.
-    final foundSnapshot = analysesNode.children.where((analysis) {
-      if (analysis is AnalysisSnapshotReference) {
-        final AnalysisSnapshotReference node = analysis;
-        return node.dateTime == currentSnapDT;
-      }
-      return false;
-    });
-
-    if (foundSnapshot.isNotEmpty && foundSnapshot.first.children.isNotEmpty) {
-      // TODO(terry): Disable Analyze button if analysis exist for current snapshot.
-      logger.log('Analysis already computed.', logger.LogLevel.warning);
+    final snapshot = controller.computeSnapshotToAnalyze;
+    final currentSnapDT = snapshot.collectedTimestamp;
+    // Analysis already completed we're done.
+    final foundMatch = controller.completedAnalyses
+        .where((element) => element.dateTime.compareTo(currentSnapDT) == 0);
```

Comment:
again should we use firstWhere here?

Reply:
Yes firstWhere and changed to only use in debug build.

```
  void _debugCheckAnalyses(DateTime currentSnapDateTime) {
    // Debug only check.
    assert(() {
      // Analysis already completed we're done.
      final foundMatch = controller.completedAnalyses.firstWhere(
        (element) => element.dateTime.compareTo(currentSnapDateTime) == 0,
        orElse: () => null,
      );
      if (foundMatch != null) {
        logger.log(
          'Analysis '
          '${MemoryController.formattedTimestamp(currentSnapDateTime)} '
          'already computed.',
          logger.LogLevel.warning,
        );
      }
      return true;
    }.call());
  }

  void _analyze() {
    final AnalysesReference analysesNode = controller.findAnalysesNode();
    assert(analysesNode != null);

    final snapshot = controller.computeSnapshotToAnalyze;
    final currentSnapDateTime = snapshot?.collectedTimestamp;

    _debugCheckAnalyses(currentSnapDateTime);
```